src/ring_theory/roots_of_unity.lean

```diff
@@ -581,6 +618,88 @@ begin
     exact ⟨i, ⟨hin, hi.symm⟩, H⟩ }
 end
 
+/-The sets `primitive_roots k R` are pairwise disjoint. -/
+lemma prim_roots_disjoint {k l : ℕ} (hk : 0 < k) (hl : 0 < l) (h : k ≠ l) :
+  disjoint (primitive_roots k R) (primitive_roots l R) :=
+begin
+  intros z hz,
+  simp only [finset.inf_eq_inter, finset.mem_inter] at hz,
+  have hkl : k = l,
+  { apply nat.dvd_antisymm,
+    { have hpowk := ((is_primitive_root.iff_def z l).1 ((mem_primitive_roots hl).1 hz.2)).1,
+      exact ((is_primitive_root.iff_def z k).1 ((mem_primitive_roots hk).1 hz.1)).2 l hpowk },
+    { have hpowl := ((is_primitive_root.iff_def z k).1 ((mem_primitive_roots hk).1 hz.1)).1,
+      exact ((is_primitive_root.iff_def z l).1 ((mem_primitive_roots hl).1 hz.2)).2 k hpowl } },
+  exfalso,
+  exact h hkl
+end
+
+/-If in `R` there is a `n`-th primitive root of unity and `b` divides `n`,
+then there is in `R` a `b`-th primitive root of unity. -/
+lemma prim_of_pow_prim {ζ : R} {n : ℕ+} {a b : ℕ} (h : is_primitive_root ζ n) (hprod : ↑n = a * b)
+  : is_primitive_root (ζ ^ a) b :=
+begin
+  apply (iff_def (ζ ^ a) b).2,
+  split,
+  { calc (ζ ^ a) ^ b
+       = ζ ^ (n : ℕ) : by rw [← pow_mul', mul_comm, hprod]
+   ... = 1 : by rw ((iff_def ζ n).1 h).1 },
+  { intros l hl,
+    rw [← pow_mul', mul_comm] at hl,
+    have hdiv := ((iff_def ζ n).1 h).2 (a * l) hl,
+    rw hprod at hdiv,
+    have hzeroa : a ≠ 0,
+    { by_contra hz,
+      simp only [not_not] at hz,
+      simp [hz] at hprod,
+      exact hprod },
+    exact (mul_dvd_mul_iff_left hzeroa).1 hdiv }
+end
+
+/-`nth_roots n` as a finset is equal to the uniun of `primitive_roots i R` for `i ∣ n`
+if there is a primitive root of unity in `R`. -/
+lemma root_of_unity_eq_uniun_prim {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) :
+  (⟨nth_roots n (1 : R) , nroots_nodup h⟩ : finset R) =
```

Comment:
Can you please explain why you don't use `roots_of_unity` here? That's what the name suggests...

Reply:
This is just because of my ignorance and I do not know how to consider `roots_of_unity n R` as a `finset`. If you can give me an hint on how to do this I will be happy to modify the rest.